libia2/include/ia2.h

```diff
@@ -226,28 +228,125 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+/* Allocate and protect the stack for this thread's i'th compartment. */
+#define ALLOCATE_COMPARTMENT_STACK(i)                                          \
+  {                                                                            \
+    extern __thread void *ia2_stackptr_##i;                                    \
+                                                                               \
+    /* We must change the pkru to write the stack pointer because each */      \
+    /* stack pointer is part of the compartment whose stack it points to. */   \
                                                                                \
-  /* Allocate a fixed-size stack. */                                           \
-  static char *allocate_stack(int i) {                                         \
+    __asm__ volatile("mov $" #i                                                \
+                     ",%%rdi\n"                                                \
+                     "push %%rbp\n"                                            \
+                     "# allocate a stack; puts its top in rax\n"               \
+                     "call allocate_stack\n"                                   \
+                     "mov %%rax, %%rbp"                                        \
+                     "# zero eax/ebx/ecx/edx\n"                                \
+                     "xor %%eax,%%eax\n"                                       \
+                     "mov %%eax,%%ebx\n"                                       \
+                     "mov %%eax,%%ecx\n"                                       \
+                     "mov %%eax,%%edx\n"                                       \
+                     "# read old pkru\n" IA2_RDPKRU                            \
+                     "\n"                                                      \
+                     "# save pkru in r12d\n"                                   \
+                     "mov %%eax,%%r12d\n"                                      \
+                     "# write new pkru"                                        \
+                     "mov_pkru_eax " #i "\n" IA2_WRPKRU                        \
+                     "\n"                                                      \
+                     "mov ia2_stackptr_" #i                                    \
+                     "@GOTTPOFF(%%rip), %%r11\n"                               \
+                     "cmpq $0x0,%%fs:(%%r11)\n"                                \
+                     "je .Lfresh_init" #i                                      \
+                     "\n"                                                      \
+                     "mov $" #i                                                \
+                     ", %%rdi\n"                                               \
+                     "call ia2_reinit_stack_err\n"                             \
```

Comment:
I would add a `ud2` after this call just to make sure it never returns, as returning would be a problem.

Reply:
Done.